### PR TITLE
+Adapter cl

### PR DESCRIPTION
**Pull request by rabeehkarimimahabadi**
Implements an adapter layer and adapter controller which chooses between the adapter layers.

### CODE REVIEW

File: seq2seq/adapters/adapter_modeling.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+
+class Adapter(nn.Module):
+  def __init__(self, model_config, adapter_config):
+    super().__init__()
+    self.input_size = model_config.d_model
+    self.add_layer_norm_after_adapter = adapter_config.add_layer_norm_after_adapter
+    self.weight_init_range = adapter_config.weight_init_range
+    # If reduction factor is not passed we consider default value of 2.
+    reduction_factor = adapter_config.reduction_factor if adapter_config.reduction_factor is not None else 2
```
Comment (jamiekiros) [inline seq2seq/adapters/adapter_modeling.py:28]: reduction_factor is already defaulted to 16 in AdapterConfig. You may also want an explicit divisibility check for input_size and reduction_factor.

**Comment by rabeehkarimimahabadi** [comment]
Hi Jamie,
Do you mind explaining a bit more what do you mean by explicit visibility
check?
thanks
Best
Rabeeh

**Review by jamiekiros — APPROVED** [review]
LGTM

File: seq2seq/adapters/adapter_controller.py [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+    for task in tasks:
+      adapter = self.get_adapter(task)
+      for param in adapter.parameters():
+        param.requires_grad = False
+
+  def convert_to_list(self, tasks):
+    if isinstance(tasks, list):
+      return tasks
+    else:
+      return [tasks]
```
Comment (georgedahl) [inline seq2seq/adapters/adapter_controller.py:71]: Can just write this as
if isinstance .... : 
  return tasks
return [tasks]

File: seq2seq/adapters/adapter_modeling.py [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+"""Implements an Adapter Layer and Meta Adapter layer."""
+import torch.nn as nn
+import torch.nn.functional as F
+from transformers.activations import get_activation
+
+from .adapter_utils import Activations, linear_layer
+
+
+class Adapter(nn.Module):
+  """Convetional Adapter layer, in which the weights of up and down sampler modules
```
Comment (georgedahl) [inline seq2seq/adapters/adapter_modeling.py:24]: "Conventional"

**Review by georgedahl — APPROVED** [review]
Sorry for the delay, just left a few minor comments.

File: seq2seq/adapters/adapter_modeling.py (hunk 1 above)
Comment (rabeehk) [reply seq2seq/adapters/adapter_modeling.py]: done, thanks

File: seq2seq/adapters/adapter_controller.py (hunk 2 above)
Comment (rabeehk) [reply seq2seq/adapters/adapter_controller.py]: done, thanks

File: seq2seq/adapters/adapter_modeling.py (hunk 3 above)
Comment (rabeehk) [reply seq2seq/adapters/adapter_modeling.py]: done, thanks

**Review by georgedahl — APPROVED** [review]

**Review by jamiekiros — APPROVED** [review]

**Comment by rabeehkarimimahabadi** [comment]
Implements adapter layers.

**Pull request merged**